### ISSUE TITLE
[DataGrid] Allow theme customizations

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1755]
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Extending the style of the data grid components in the theme creates TypeScript errors.

## Expected Behavior 🤔

No TypeScript errors when customizing the data grid.

## Steps to Reproduce 🕹

Steps:

1. Open https://codesandbox.io/s/material-demo-forked-2l2iu?file=/demo.tsx

```jsx
const theme = createMuiTheme({
  overrides: {
    MuiDataGrid: {
      root: {
        backgroundColor: "red"
      }
    }
  }
});
```

2. See the error 💥

<img width="520" alt="Screenshot 2021-05-24 at 12 13 55" src="https://user-images.githubusercontent.com/3165635/119332926-8b631500-bc89-11eb-9395-707aafe57d23.png">

## Context 🔦

Reported by @LeoLozes in https://github.com/mui-org/material-ui/issues/19427#issuecomment-846828746.

## Your Environment 🌎

v4.0.0-alpha.29

**Comment by oliviertassinari** [comment]
For the solution, we can reproduce the solution used by `@material-ui/lab` to extend the core: https://next.material-ui.com/components/about-the-lab/#typescript. I think that it would also be good to explore how to make this built-in.

**Comment by Primajin** [comment]
Hey there - I'm having a question regarding the overrides - are they actually read besides root? Putting the TypeError aside I am trying to use global overrides rather than using local classes and assign them to XGrid. (might be a duplicate of #919)

```json
    cell: {
      background: 'red !important',
      content: 'hello!',
      padding: 0,
    },
    root: {
      borderColor,
      borderRadius: 0,
      borderWidth: '1px 0 0',
    },
```
while I can see the styles from root applied - I can not see the ones from cell applied 🤔

**Comment by oliviertassinari** [comment]
@Primajin Thanks for raising the duplicate. I have closed it. This is a simple problem to fix. But I think that we should wait for depending on v5 only to have a built-in fix. Developers can extend the theme manually for now, as long as we export the right type, which we could add now.

Regarding the available classes, you only have the `root` available. 
https://github.com/mui-org/material-ui-x/blob/14a2b22b371579778bfd06539b8807dc0314b2e1/packages/grid/_modules_/grid/components/containers/GridRootStyles.ts#L14
The docs is wrong https://material-ui.com/api/data-grid/data-grid/#css. We broke it in #1694. @ZeeshanTamboli could you revert?

**Comment by ZeeshanTamboli** [comment]
@oliviertassinari What about the `classes` prop docs from PR #1660? Since [GridClasses](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/models/gridClasses.ts) interface is not yet documented will it be fine?

**Comment by oliviertassinari** [comment]
@ZeeshanTamboli This is not the same type

**Comment by ZeeshanTamboli** [comment]
@oliviertassinari 

> The classes API should equal the theme.components.DataGrid.styleOverrides API,

So should this be fixed for v4 currently?

> what should the docs ## CSS section cover, IMHO, the public class name API

What is public class name API? Is it `className`?

> we don't encourage so much to use classes in v5.

I believe we are still not fully dependent on core v5. So should we still remove documentation for v4 users using useStyles?

So I think we need to decide between v4 or v5. I'd say keep it until v5 is fully released and users adopt it. Until then, we could document that it is not supported in global `overrides` but only for `makeStyles` classes API.

**Comment by Primajin** [comment]
If it helps I was able to make it work via overrides --> root --> and then using string classes:

```js
    root: {
      '& .MuiDataGrid-cell': {
        borderBottomColor: borderColor,
        textOverflow: 'ellipsis',
        whiteSpace: 'pre', // don't collapse double spaces
      },
      '& .MuiDataGrid-cell:focus-within, & .MuiDataGrid-colCell:focus-within,  & .MuiDataGrid-columnHeader:focus-within':
        {
          outline: 0,
        },
      '& .MuiDataGrid-columnsContainer': {
        borderBottomColor: borderColor,
      },
      '& .MuiDataGrid-columnSeparator--resizable': {
        color: borderColor,
      },
      '& .MuiDataGrid-row.focusedNotSelected': {
        backgroundColor: focusColor,
      },
      borderColor,
      borderRadius: 0,
      borderWidth: '1px 0 0',
    },
```

**Comment by oliviertassinari** [comment]
@Primajin This is very close to how we want to encourage customizations. In #1950, we will expose a `dataGridClasses` so you can get type safety when you do your customizations.

@ZeeshanTamboli Actually, I think that we can keep the documentation as is. It's compatible with v5. So what do we have next? We have two independent problems to fix:

1. [ ] Fix the TypeScript theme module augmentation, we can do it now
2. [ ] Fix https://github.com/mui-org/material-ui-x/issues/1755, we need to wait for v5 only support, which we target for September 1st.

**Comment by ZeeshanTamboli** [comment]
This issue is now solved on latest version `5.0.0-beta.1`.

CodeSandbox for the same: https://codesandbox.io/s/datagriddemo-material-demo-forked-8jryd?file=/demo.js

**Comment by justdvl** [comment]
Thank you, this works great.
